[Dependency Updates] Test Instructions Documentations per Dependency Update

ParaskP7 opened the pull request:
Parent #17551
Internal Discussion `pcdRpT-2Hp-p2`

This PR adds a new documentations page with [test instructions per dependency update](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md).

FYI: The goal is to further automate the dependency update process by documenting all the specific dependency updates that require explicit testing instructions. The hypothesis here is that this approach would not only save time and effort for each update, but also elevate the quality of dependency update work by ensuring that the testing aspect is not left to chance.

PS: This PR is also slightly updating the [pull request template](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/.github/PULL_REQUEST_TEMPLATE.md) by:
1. Splitting each section using a horizontal line, for grouping purposes (38a4d69483d544015bc05df5fb92cbfcd0f62318),
2. Starting each section with a second-level heading, for consistency purposes (6b80fa7282112eb518479a3fe89e7d92b90d128b), and
3. Applying other minor changes, for consistency purposes (273a96f21311094292777482ad257224c194ca83).

-----

@irfano I added you as the main reviewer, not so randomly (you shared lots of good ideas), since I just wanted someone from the Jetpack/WordPress mobile team to be aware and sign-off on that change. Feel free to merge this PR directly yourself if you deem so.

-----

## To Test:

1. Proofread the new [test instructions per dependency update](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md) documentation page.
    - [Purpose](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md#purpose)
    - [FAQ](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md#faq)
    - [Content](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md#content)
    - [Smoke Test](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md#smoke-test)
    - [Special](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md#special)
2. Verify that each section is well written, can be easily understood and provide enough value while devs are working on a specific dependency update.

FYI: Note that the [Smoke Test](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md#smoke-test) and [Special](https://github.com/wordpress-mobile/WordPress-Android/blob/docs/test_instructions_per_dependency_update/docs/test_instructions_per_dependency_update.md#special) sections can be further improved. However, I didn't want make this improvement now and instead added TODO for devs to keep improving the documentation themselves when and if needed.

-----

## Regression Notes

1. Potential unintended areas of impact

    - N/A

4. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To Test` section above.

5. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

ParaskP7 committed:
Docs: Add new test instructions per dependency update wiki page

ParaskP7 committed:
Docs: Link test instructions per dependency update into pr template

ParaskP7 committed:
Docs: Split pr template per section using a horizontal line

ParaskP7 committed:
Docs: Start each pr template section with a second-level heading

ParaskP7 committed:
Docs: Capitalize each pr template section

ParaskP7 committed:
Docs: Add todo bullet-points on regression notes pr template section

ParaskP7 committed:
Docs: Add a faq section to test instructions per dep update wiki page

ParaskP7 committed:
Docs: Add jp/wp test instruction prefix per test scenario

ParaskP7 committed:
Docs: Add two additional test scenarios on the work manager dependency

TODOs got added on both those test scenarios as I am not sure how
to test this functionality and can't easily find the entry point to it.
All I could do to test this was to manually comment out some code and
verify that the functions were called after setting the device's date to
days/weeks after today.

Still, I think it is better to have that there and inform future
dependency update testers about this functionality and the fact that
this needs to be (somehow) tested, rather than ignore both completely.

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/17892

ParaskP7 committed:
Docs: Remove unnecessary test scenario from the preference dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/17960

ParaskP7 committed:
Docs: Update test scenarios on the preference dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/17960

ParaskP7 committed:
Docs: Add test scenarios for the exoplayer dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/17936

ParaskP7 committed:
Docs: Add test scenarios for the firebase dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/18552

ParaskP7 committed:
Docs: Add test scenarios for the play services auth dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/17994

ParaskP7 committed:
Docs: Add test scenarios for the play services code scanner dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/17996

ParaskP7 committed:
Docs: Add test scenarios for the navigation dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/18554

ParaskP7 committed:
Docs: Add test scenarios for the navigation dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/18113

ParaskP7 committed:
Docs: Add test scenarios for the zendesk dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/18090

ParaskP7 committed:
Docs: Add test scenarios for the jsoup dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/18102

ParaskP7 committed:
Docs: Add test scenarios for the auto service and kotlin poet deps

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/18076

ParaskP7 committed:
Docs: Add test scenarios for the lottie dependency

TODOs got added on all those test scenarios. I think it is better to
have that there and inform future dependency update testers about this
functionality and the fact that this needs to be (somehow) tested,
rather than ignore both completely.

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/17571

ParaskP7 committed:
Docs: Add test scenarios for the ucrop dependency

For more info see: https://github.com/wordpress-mobile/
WordPress-Android/pull/18060

ParaskP7 committed:
Docs: Split dependency update test instructions using a horizontal line

ParaskP7 committed:
Docs: Group content of dependency updates into logical units

ParaskP7 committed:
Docs: Move plugin group content of dependency updates on top

ParaskP7 committed:
Docs: Add 3 more test scenarios for webkit, firebase, mlkit dependencies

TODOs got added on all those test scenarios. I think it is better to
have that there and inform future dependency update testers about this
functionality and the fact that this needs to be (somehow) tested,
rather than ignore both completely.

ParaskP7 committed:
Docs: Add smoke test section

FYI: Some smoke test instructions might have an 'Extra' section with
addition instructions.

ParaskP7 committed:
Docs: Add special section

FYI: Every special test instructions have a 'Why & Extra' section to
further explain why that dependency update is so special and how to
deal with it.

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+
+-----
+
+### WorkManager [[androidxWorkManagerVersion](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/build.gradle#L54)] <a name="workmanager"></a>
```
All other links point to the first line (`#L1`), but the `androidxWorkManagerVersion` and `androidxPreferenceVersion` links point to `#L54` and `#L50`. I believe you avoided stating line numbers in case they change in the future. We can remove all line numbers from the links and just link the file.

ParaskP7 replied:
> All other links point to the first line (#L1), but the androidxWorkManagerVersion and androidxPreferenceVersion links point to #L54 and #L50.

Spot on! 💯 

> I believe you avoided stating line numbers in case they change in the future.

Exactly! 🎯 

> We can remove all line numbers from the links and just link the file.

True, will do exactly that! 🥇

ParaskP7 replied:
Done: a6f1cd00b94de220c46607e8897ed8159e850e9b + 847d28b32c5632bd954ea96f0ddef35811c71530

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+### Navigation [[navigationVersion](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/settings.gradle#L1)] <a name="navigation"></a>
+
+<details>
+    <summary>1. [JP/WP] Image Editing Flow [libs:image-editor]</summary>
+
+- Add a new `blog` post.
+- Add a new `image` block.
+- Choose an image and wait for it to be uploaded within the `image` block.
+- Click on the `media options` of this image (top right) and then click `edit`.
+- Verify that the `Edit Image` screen is shown and functioning as expected.
+- Crop the image and click the `done` menu option (top right).
+- Make sure the image is updated accordingly.
+
```
This section is "Navigation". We should add testing for navigation dependency instead image-editor.

ParaskP7 replied:
Ah @irfano , this is an ad-hoc convention I used to help the reader locate where that dependency and it's testing scenario is being used, just like I am doing for example on the `Preference` testing instructions, where I point to an exact file instead, see `1. [JP/WP] Account Settings [AccountSettingsFragment.kt]` and notice the `[AccountSettingsFragment.kt]` suffix.

In this case, what I want to convey here is that the `Navigation` dependency is only being used by this module, the `[libs:image-editor]`, and thus this needs to be the focus of this testing scenario.

Does that make sense. Do you have any idea on how to do that better? 🤔 

Else, would you prefer to remove this information from the testing scenarios altogether, wdyt? 🤔

irfano replied:
> the Navigation dependency is only being used by this module, the [libs:image-editor]

That totally makes sense! 👍🏻

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+
+-----
+
+### Preference [[androidxPreferenceVersion](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/build.gradle#L50)] <a name="preference"></a>
```
All the screens in this section (except `NotificationsSettingsFragment`) use `android.preference` library instead of the `androidx.preference`. I believe we should remove all of them except `NotificationsSettingsFragment`

ParaskP7 replied:
Yea, you're right about that @irfano and I was wondering how to go best about it, just because of the pending `android.preference` to `androidx.preference` migration... 🤔 

What about instead of removing the how-to-test section and thus all those testing scenarios, to instead add an higher level info within that `Preference` section explaining that fact, including links to the migration issue, `android.preference` to `androidx.preference`, but still keep those testing scenarios, wdyt? 🤔

irfano replied:
I would prefer to remove those screens because we don't know when to migrate to `androidx.preference`. Screens and features may change until we migrate, and it could be pointless to keep these test cases until we migrate. However, I understand your point. You think that these cases are already written and do not lose them. 
So, I leave the decision to you; both options are fine with me.

ParaskP7 replied:
👋 @irfano , wow about this 2f1403d9a7efc9e3cde2a4ff6e53815e789c7d19 change that I pushed just now? 🤔

irfano replied:
Better now. 😊 Thanks for the change.

ParaskP7 replied:
Awesome, thank YOU for helping me doing this change, YOU rock! 🤟

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+
+### Firebase [[firebaseBomVersion](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/build.gradle#L1)] <a name="firebase"></a>
+
+<details>
+    <summary>1. [JP/WP] Push Notification on Post Being Published [GCMMessageService.java]</summary>
+
+- Add a new `blog` post.
+- Add any number of blocks.
+- Publish the post.
+- Verify that, on post being published, you get a push notification with the post's title, which,
+  when you click on it, navigates you to the 'Notifications Detail' screen for this post.
```
I don't receive a push notification after publishing a post. I saw a notification that shows the publishing is in progress, but it must be a local notification instead of a push notification. I think the better case to test `GCMMessageService` could be commenting on a post from the web and receiving a mobile push notification for it.

ParaskP7 replied:
Actually @irfano , the reason I am confident that this is indeed a push notification is because when I do that, and having multiple Jetpack apps installed on my test device, including the production, beta and alpha versions, I get a notification on all 3 apps, including on my iOS test device and the app installed there. 🤔 

Maybe you just need to enable the notifications on your testing device/app by clicking `Fix` on the `Notifications` tab? 🤔

irfano replied:
I've already enabled the notification permission.
Are we talking about the same flow? I am sharing the video of what I am doing and not receiving notification.

[Screen_recording_20231110_163640.webm](https://github.com/wordpress-mobile/WordPress-Android/assets/2471769/efd99639-8561-4173-800d-670040735752)

ParaskP7 replied:
Yes @irfano , as far as I understand we are talking about the same flow... 🤷 

Please check my recording below:

https://github.com/wordpress-mobile/WordPress-Android/assets/9729923/89a9b70e-1420-4379-b2c2-d59c57e26a40

I might be missing something here... 🤔

irfano replied:
Okay! I have received the notification now with a different account. I believe there are certain criteria on the backend. Even if we use a different test case here, there may still be some criteria for that case. So, let's keep this flow. 👍🏻

ParaskP7 replied:
> Okay! I have received the notification now with a different account.

🎉 

> I believe there are certain criteria on the backend.

Oh, this is a TIL for me that there are some criteria to received push notification! 😮 

> Even if we use a different test case here, there may still be some criteria for that case. So, let's keep this flow. 👍🏻

True, and thanks again for being so thorough @irfano ! 🙇

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+<details>
+  <summary>2. [WP] Google Sign-in on WordPress</summary>
+
+ℹ️ Prerequisite: If you don't have access to upload an`.aab` to Google Play Store for
```
Need a white space after `...to upload an`.

ParaskP7 replied:
Wow, nice catch! 🥇

ParaskP7 replied:
Done: c243351f35561490525a9f791701670515ce6330

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+  <summary>2. [WP] Google Sign-in on WordPress</summary>
+
+ℹ️ Prerequisite: If you don't have access to upload an`.aab` to Google Play Store for
+   `WordPress - Website Builder`, post a request to `+systemsrequests`.
```
I think we should mention that this section is not for external contributors.

ParaskP7 replied:
True, let me do that! 🥇

ParaskP7 replied:
Done: 2de223d49a164bb2c8689ff1b4cd3364a5fadd0a

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+<details>
+  <summary>1. [JP] Google Sign-in on Jetpack</summary>
+
+ℹ️ Prerequisite: If you don't have access to upload an`.aab` to Google Play Store for
```
Need a white space after `...to upload an`.

ParaskP7 replied:
[SAME](https://github.com/wordpress-mobile/WordPress-Android/pull/19552#discussion_r1389266457)

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+Step.1:
+- Build and install the `Jetpack` app (note that you don't need a release build, a debug build will
+  suffice).
+- Login to the `Jetpack` app with a `WP.com` account (note that you need to use a non `A8C` account
```
np: `Login to the` → `Log in to the`
Since "Log in" is the verb.

ParaskP7 replied:
Love the detail of your review! 🥇

ParaskP7 replied:
Done: 84bf847dbb27fd4ee2e2a24b1fb1b54aa29740ab

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+- Head back to your mobile.
+- Tap the `Scan Login Code` item on the `Me` screen you are currently at.
+- Scan the QR code on the web browser.
+- Follow the remaining prompts on your mobile to login to WordPress on your web browser (desktop),
```
np: `to login to WordPress` → `to log in to WordPress`
Since "Log in" is the verb.

ParaskP7 replied:
[SAME](https://github.com/wordpress-mobile/WordPress-Android/pull/19552#discussion_r1389271813)

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+  (or non if you don't have any yet).
+- Go back to the `Help` settings screen and then click on `Contact Support`.
+- Verify that the `Help` Zendesk related screen opens and that you are prompt to `Send a message...`.
+- Type `Testing Zendesk X.Y.Z update, please ignore and thank you!` and then send your message.
```
This step is outdated now. The `Contact support` button opens the chatbot. We need to tap the "Contact support" button after sending the first message to the bot.

ParaskP7 replied:
Awesome @irfano , thanks for that correction, I will add an extra step to tap on the `Contact support 📢` button, tested it and it  indeed works as you described, not quite outdated, just enhanced let me say. 😊

ParaskP7 replied:
Done: b1dd15b749c0b8fbf791814611b040dff373c5ca

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,859 @@
+<details>
+    <summary>1. [JP/WP] Classic Editor [Aztec]</summary>
+
+ℹ️ Disable the `Block Editor` before testing this.
```
Can we add steps for disabling and enabling `Block Editor`?

ParaskP7 replied:
Good idea! 💯

ParaskP7 replied:
Done: f3870ce23a6e1a3587dab105bb199da5cb7b2b3e

irfano reviewed (changes requested):
Thank you for your efforts on this. I've added some comments for your review.

ParaskP7 committed:
Docs: Point to build gradle line one instead of a specific line number

This is done for 'WorkManager' and 'Preference', which were both
incorrectly pointing to a specific line number, instead of the default
line one.

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1388166635

ParaskP7 committed:
Docs: Remove unnecessary line numbers from links and point to file

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1388166635

ParaskP7 committed:
Docs: Add missing white spaces on play services auth test instructions

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1388214910

ParaskP7 committed:
Docs: Add info message about external contributor on play services auth

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1388217235

ParaskP7 committed:
Docs: Change login wording to log-in on play services code scanner

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1388224370

ParaskP7 committed:
Docs: Add an extra testing step on zendesk

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1388231514

ParaskP7 committed:
Docs: Add an extra info about disabling/enabling block editor on jsoup

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1388239341

irfano commented on docs/test_instructions_per_dependency_update.md:
```diff
@@ -0,0 +1,870 @@
+
+-----
+
+### GoogleServices [[googleServicesVersion](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/settings.gradle#1)] <a name="googleservices"></a>
+
+-----
+
+### Dagger [[daggerVersion](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/settings.gradle#1)] <a name="dagger"></a>
+
+-----
+
+### Kotlin [[kotlinVersion](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/settings.gradle#1)] + [[androidxComposeCompilerVersion](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/build.gradle)] <a name="kotlin"></a>
```
np: `googleServicesVersion`, `daggerVersion`, and `kotlinVersion` links still contain `#1`. We can remove them and link `https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/settings.gradle` file for consistency.

ParaskP7 replied:
OMG, sorry for that @irfano and thanks for being so thorough here, I replaced all usages of `#L1`, but I also had `#1` there, whot... 🤷

ParaskP7 replied:
Done: 10796939f0c60174d453163661bf6e28fd1b350c

ParaskP7 committed:
Docs: Remove remaining line numbers from links and point to file

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1389360965

ParaskP7 committed:
Docs: Split preference into androidx and android testing instructions

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19552#discussion_r1389373449

irfano reviewed (approved):
Thank you for submitting this PR and addressing my comments.
LGTM. 👍🏻

ParaskP7 commented:
Awesome, thank you so much reviewing and merging this, more so for being as thorough as you were, you rock! 🙇 ❤️ 🚀
